`MatrixMockProjectFactory` should be an optional extension

```
... hudson.ExtensionFinder$Sezpoz scout
WARNING: Failed to scout jenkins.plugin.mockloadbuilder.MatrixMockProjectFactory
java.lang.NoClassDefFoundError: hudson/matrix/MatrixProject
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:348)
	at hudson.ExtensionFinder$Sezpoz.scout(ExtensionFinder.java:669)
	at hudson.ClassicPluginStrategy.findComponents(ClassicPluginStrategy.java:473)
	at hudson.ExtensionList.load(ExtensionList.java:365)
	at hudson.ExtensionList.ensureLoaded(ExtensionList.java:303)
	at hudson.ExtensionList.getComponents(ExtensionList.java:168)
	at jenkins.model.Jenkins$8.onInitMilestoneAttained(Jenkins.java:1104)
	at jenkins.InitReactorRunner$1.onAttained(InitReactorRunner.java:82)
	at org.jvnet.hudson.reactor.ReactorListener$Aggregator.onAttained(ReactorListener.java:104)
	at org.jvnet.hudson.reactor.Reactor$1.run(Reactor.java:176)
	at org.jvnet.hudson.reactor.Reactor$Node.run(Reactor.java:117)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.ClassNotFoundException: hudson.matrix.MatrixProject
	at jenkins.util.AntClassLoader.findClassInComponents(AntClassLoader.java:1374)
	at jenkins.util.AntClassLoader.findClass(AntClassLoader.java:1327)
	at jenkins.util.AntClassLoader.loadClass(AntClassLoader.java:1080)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	... 15 more
```

`optional=true` would work iff the class signature also required some type from `matrix-project`, most easily by changing `MockProjectFactory` to have a job type parameter and making `create` return `MatrixProject`.

(refiled from [JENKINS-47362](https://issues.jenkins.io/browse/JENKINS-47362))

--- a/pom.xml
+++ b/pom.xml
@@ -87,6 +87,10 @@
     </dependency>
     <dependency>
       <groupId>org.jenkins-ci.plugins</groupId>
+      <artifactId>variant</artifactId>
+    </dependency>
+    <dependency>
+      <groupId>org.jenkins-ci.plugins</groupId>
       <artifactId>junit</artifactId>
     </dependency>
     <dependency>
@@ -107,6 +111,11 @@
     <dependency>
       <groupId>org.jenkins-ci.plugins.workflow</groupId>
       <artifactId>workflow-basic-steps</artifactId>
+      <optional>true</optional>
+    </dependency>
+    <dependency>
+      <groupId>org.jenkins-ci.plugins.workflow</groupId>
+      <artifactId>workflow-durable-task-step</artifactId>
       <optional>true</optional>
     </dependency>
     <dependency>

--- a/src/main/java/jenkins/plugin/mockloadbuilder/CreateMockLoadMatrix.java
+++ b/src/main/java/jenkins/plugin/mockloadbuilder/CreateMockLoadMatrix.java
@@ -1,8 +1,8 @@
 package jenkins.plugin.mockloadbuilder;
 
-import hudson.Extension;
+import org.jenkinsci.plugins.variant.OptionalExtension;
 
-@Extension(optional = true)
+@OptionalExtension(requirePlugins = "matrix-project")
 public class CreateMockLoadMatrix extends AbstractCreateMockLoad {
     @Override
     public String getShortDescription() {

--- a/src/main/java/jenkins/plugin/mockloadbuilder/CreateMockLoadPipeline.java
+++ b/src/main/java/jenkins/plugin/mockloadbuilder/CreateMockLoadPipeline.java
@@ -1,8 +1,8 @@
 package jenkins.plugin.mockloadbuilder;
 
-import hudson.Extension;
+import org.jenkinsci.plugins.variant.OptionalExtension;
 
-@Extension(optional = true)
+@OptionalExtension(requirePlugins = {"workflow-job", "workflow-cps", "workflow-basic-steps", "workflow-durable-task-step"})
 public class CreateMockLoadPipeline extends AbstractCreateMockLoad {
     @Override
     public String getShortDescription() {

--- a/src/main/java/jenkins/plugin/mockloadbuilder/MockLoadStep.java
+++ b/src/main/java/jenkins/plugin/mockloadbuilder/MockLoadStep.java
@@ -1,7 +1,6 @@
 package jenkins.plugin.mockloadbuilder;
 
 import com.google.common.collect.ImmutableSet;
-import hudson.Extension;
 import hudson.FilePath;
 import hudson.Launcher;
 import hudson.model.TaskListener;
@@ -16,6 +15,7 @@ import org.kohsuke.stapler.DataBoundConstructor;
 import java.io.InputStream;
 import java.io.OutputStream;
 import java.util.Set;
+import org.jenkinsci.plugins.variant.OptionalExtension;
 
 public class MockLoadStep extends Step {
 
@@ -34,7 +34,7 @@ public class MockLoadStep extends Step {
         return new Execution(this, context);
     }
 
-    @Extension(optional = true)
+    @OptionalExtension(requirePlugins = "workflow-step-api")
     public static class DescriptorImpl extends StepDescriptor {
         @Override
         public Set<? extends Class<?>> getRequiredContext() {

--- a/src/main/java/jenkins/plugin/mockloadbuilder/PipelineMockProjectFactory.java
+++ b/src/main/java/jenkins/plugin/mockloadbuilder/PipelineMockProjectFactory.java
@@ -1,16 +1,15 @@
 package jenkins.plugin.mockloadbuilder;
 
-import hudson.Extension;
 import hudson.model.Job;
-import hudson.tasks.LogRotator;
 import jenkins.model.Jenkins;
 import jenkins.model.ModifiableTopLevelItemGroup;
 import org.jenkinsci.plugins.workflow.cps.CpsFlowDefinition;
 import org.jenkinsci.plugins.workflow.job.WorkflowJob;
 
 import java.io.IOException;
+import org.jenkinsci.plugins.variant.OptionalExtension;
 
-@Extension(optional = true)
+@OptionalExtension(requirePlugins = {"workflow-job", "workflow-cps", "workflow-basic-steps", "workflow-durable-task-step"})
 public class PipelineMockProjectFactory extends MockProjectFactory {
     @Override
     public int getFrequency() {

